workflows/sanbi_cwltutorial/README.md

```diff
@@ -0,0 +1,167 @@
+        in:
+          infile: infile
+        out: [outfile]
+      complement:
+        run: complement.cwl
+        in:
+          infile: reverse/outfile
+        out: [outfile]
+
+Note that this is `class: Workflow` instead of `class: CommandLineTool`. It has
+its own `inputs` and `outputs`, just like a command line tool. Instead of
+linking to a command (with `baseCommand`), a workflow has `steps`, which in this
+case each refer to a CWL command line tool description file. In each step the
+inputs provided and outputs used are mentioned using `in` and `out` respectively.
+The `infile` input from `inputs` is bound to the `infile` parameter of `reverse.cwl`.
+The `outfile` output of this tool is captured and then bound to the `infile`
+parameter of `complement.cwl`. Finally `outfile` from the complement step is
+bound (using `outputSource`) to the `outfile` output of the workflow as a whole.
+
+At this (workflow) level the implementation details of the steps are hidden.
+The workflow has no need to know about the structure of the command line or
+Docker containers. It only needs to know steps and how steps relate to each other.
+
+One effect of how the steps are written is that the final output file will
+inherit patterns from each step. So `dna.txt` becomes `dna_reversed_complement.txt`.
+
+The final example `revcomp_with_rename.cwl` illustrates some of the more
+advanced features of CWL. Firstly this workflow takes an option string
+parameter (`outfile`) which is the name to give the output file of the workflow.
+This parameter is optional because its type is `string?`. Like in
+regular expressions, the `?` denotes that this paramter does not have to be
+provided. If the parameter is not provided, the output filename would be the same
+as for `revcomp.cwl`.
+
+The second difference from `revcomp.cwl` is the step
+that allows output renaming. Instead of being provided as a command line tool,
+it runs a Javascript expression. This require a `InlineJavascriptRequirement`
+to be specified and then a step of type `ExpressionTool`:
+
+    rename:
+      run:
+        class: ExpressionTool
+        inputs:
+          infile:
+            type: File
+          outfile:
+            type: string?
+        outputs:
+          outfile: File
+        expression: >
+          ${
+          var outfile = inputs.infile;
+          if (inputs.outfile) {
+            outfile.basename = inputs.outfile;
+          }
+          return { outfile: outfile }; }
+      in:
+        infile: complement/outfile
+        outfile: outfile
+      out: [outfile]
+
+This also has an optional `outfile` parameter. The Javascript tests if this is
+set and if so, uses it to modify the `outfile.basename`, resulting in the
+output file having the name specified by the user. The Javascript expression
+returns a dictionary whose keys correspond to the names of outputs specified
+for the step.
+
+The `InlineJavascriptRequirement` requires the use of NodeJS (the `node`)
```

Comment:
This paragraph is true for the reference implementation, bu other implementations might use some other Javascript/ECMAScript 5.1 environment to evaluate an Expression.